native/dlaqr4.go

```diff
@@ -0,0 +1,467 @@
+	// sufficient workspace.
+	nsmax := min((n+6)/9, 2*lwork/3)
+	nsmax -= nsmax % 2
+
+	var (
+		ndec int     // Deflation window size decrement.
+		ndfl = 1     // Number of iteration since last deflation.
+		nw   = nwmax // Start with maximum deflation window size.
+
+		itmax = max(30, 2*kexsh) * max(10, (ihi-ilo+1))
+		it    = 0
+	)
+	// Main loop.
+	for kbot := ihi; kbot >= ilo; {
+		if it == itmax {
+			unconverged = kbot + 1
+			break
+		}
+		it++
+
+		// Locate active block.
+		ktop := ilo
+		for k := kbot; k >= ilo+1; k-- {
+			if h[k*ldh+k-1] == 0 {
+				ktop = k
+				break
+			}
+		}
+
+		// Select deflation window size nw.
+		//
+		// Typical Case:
+		//  If possible and advisable, nibble the entire active block.
+		//  If not, use size min(nwr,nwmax) or min(nwr+1,nwmax)
+		//  depending upon which has the smaller corresponding
+		//  subdiagonal entry (a heuristic).
+		//
+		// Exceptional Case:
+		//  If there have been no deflations in kexnw or more
+		//  iterations, then vary the deflation window size. At first,
+		//  because, larger windows are, in general, more powerful than
+		//  smaller ones, rapidly increase the window to the maximum
+		//  possible. Then, gradually reduce the window size.
+		nh := kbot - ktop + 1
+		nwupbd := min(nh, nwmax)
+		if ndfl < kexnw {
+			nw = min(nwupbd, nwr)
+		} else {
+			nw = min(nwupbd, 2*nw)
+		}
+		if nw < nwmax {
+			if nw >= nh-1 {
+				nw = nh
+			} else {
+				kwtop := kbot - nw + 1
+				if math.Abs(h[kwtop*ldh+kwtop-1]) > math.Abs(h[(kwtop-1)*ldh+kwtop-2]) {
+					nw++
+				}
+			}
+		}
+		if ndfl < kexnw {
+			ndec = -1
+		} else if ndec >= 0 || nw >= nwupbd {
+			ndec++
+			if nw-ndec < 2 {
+				ndec = 0
+			}
+			nw -= ndec
+		}
+
+		// Split workspace under the subdiagonal of H into:
+		//  - an nw×nw work array V in the lower left-hand corner,
+		//  - an nw×nho horizontal work array along the bottom edge (nho
+		//    must be at least nw but more is better),
+		//  - an nve×nw vertical work array along the left-hand-edge
+		//    (nve can be any positive integer but more is better).
+		kv := n - nw
+		kt := nw
+		nho := n - (nw + 1) - kt
+		kwv := nw + 1
+		nve := n - nw - kwv
+
+		// Aggressive early deflation.
+		ls, ld := impl.Dlaqr2(wantt, wantz, n, ktop, kbot, nw,
+			h, ldh, iloz, ihiz, z, ldz, wr[:kbot+1], wi[:kbot+1],
+			h[kv*ldh:], ldh, nho, h[kv*ldh+kt:], ldh, nve, h[kwv*ldh:], ldh, work, lwork)
+
+		// Adjust kbot accounting for new deflations.
+		kbot -= ld
+		// ks points to the shifts.
+		ks := kbot - ls + 1
+
+		// Skip an expensive QR sweep if there is a (partly heuristic)
+		// reason to expect that many eigenvalues will deflate without
+		// it. Here, the QR sweep is skipped if many eigenvalues have
+		// just been deflated or if the remaining active block is small.
+		if ld > 0 && (100*ld > nw*nibble || kbot-ktop+1 <= min(nmin, nwmax)) {
+			// ld is positive, note progress.
+			ndfl = 1
```

Comment:
This is correct. Here we have just deflated more than nibble% of the deflation window, so we reset the counter of "iterations without deflation". But thinking about it, it seems like 0 would be a more correct value. I don't know why fortran uses 1.